ocaml/xapi/quicktest_cbt.ml

```diff
@@ -0,0 +1,153 @@
+      try
+        start enable_cbt_test;
+        test_assert ~test:enable_cbt_test
+          (not (VDI.get_cbt_enabled ~session_id ~rpc:!rpc ~self:vDI))
+          ~msg:"VDI.cbt_enabled field should be set to false for new VDIs, but wasn't ";
+        VDI.enable_cbt ~session_id ~rpc:!rpc ~self:vDI;
+        test_assert ~test:enable_cbt_test
+          (get_cbt_status vDI)
+          ~msg:"VDI.enable_cbt failed";
+        VDI.disable_cbt ~session_id ~rpc:!rpc ~self:vDI;
+        test_assert ~test:enable_cbt_test
+          (not (get_cbt_status vDI)) (* disable_cbt fails *)
+          ~msg:"VDI.disable_CBT failed";
+        success enable_cbt_test
+      with
+      | Test_assertion_failed msg -> failed enable_cbt_test msg
+      | e -> report_failure e "enable/disable CBT" enable_cbt_test in
+
+
+    (* Test data_destroy and snapshot update the necessary fields *)
+    let vdi_data_destroy_test ~vDI =
+      let data_destroy_test = make_test "Testing VDI.data_destroy" 4 in
+      try
+        start data_destroy_test;
+        VDI.enable_cbt ~session_id ~rpc:!rpc ~self:vDI;
+        debug data_destroy_test "Enabled CBT on original VDI";
+        test_assert ~test:data_destroy_test
+          (get_cbt_status vDI)
+          ~msg:"VDI.enable_cbt failed";
+        let newvdi = VDI.snapshot ~session_id ~rpc:!rpc ~vdi:vDI ~driver_params:[] in
+        debug data_destroy_test "Snapshotted original VDI.";
+        test_assert ~test:data_destroy_test
+          (get_cbt_status newvdi)
+          ~msg:"VDI.snapshot failed, cbt_enabled field didn't carry over";
+        VDI.data_destroy ~session_id ~rpc:!rpc ~self:newvdi;
+        debug data_destroy_test "Destroy snapshot VDI data";
+        test_compare ~test:data_destroy_test
+          (VDI.get_type ~session_id ~rpc:!rpc ~self:newvdi)
+          `cbt_metadata
+          ~msg:"VDI.data_destroy failed to update VDI.type";
+        let content_id_str = "/No content: this is a cbt_metadata VDI/" in
+        test_compare ~test:data_destroy_test
+          (VDI.get_other_config ~session_id ~rpc:!rpc ~self:newvdi |> List.assoc "content_id")
+          content_id_str
+          ~msg:"VDI.data_destroy failed to update VDI.content_id";
+        VDI.destroy ~session_id ~rpc:!rpc ~self:newvdi;
+        debug data_destroy_test "Snapshot destroyed.";
+        success data_destroy_test
+      with
+      | Test_assertion_failed msg -> failed data_destroy_test msg
+      | e -> report_failure e "data_destroy" data_destroy_test in
+
+
+    let run_test_suite ~sR ~vDI =
+      let sr_ops = (SR.get_allowed_operations ~session_id ~rpc:!rpc ~self:sR) in
+      [ (fun () -> enable_disable_cbt_test ~vDI) ,
+        [ `vdi_enable_cbt ; `vdi_disable_cbt ]
+      ; (fun () -> vdi_data_destroy_test ~vDI) ,
+        [ `vdi_enable_cbt ; `vdi_data_destroy ; `vdi_snapshot ; `vdi_destroy ]
+      ]
+      |> List.iter
+        (fun (test,list_vdi_ops) ->
+           if List.for_all (fun vdi_op -> List.mem vdi_op sr_ops) list_vdi_ops
```

Comment:
What if `List.for_all` returns false and the test is not run? Is this a reason for concern? Right now this simply ignored. In the worst case, `run_test_suite` would do nothing.

Reply:
Hmm, the behaviour of the existing storage quicktests is exactly the same. Maybe we could print a warning though. I think this behaviour is fine, because it makes the test general, it will work on all setups - if we don't currently have an SR that's suitable for our test, that's not really a failure.
But actually I agree that a warning message would be useful.

Reply:
Considered the output of our tests, I think nobody would be able to read the warning :(

Reply:
Quicktest output is quite clean, literally just prints test names and statuses plus any given debug info. At present I've added a "skipping test as SR cannot support it" line. Worst case scenario: run_test_suite prints out a load of these and then concludes.

Reply:
We have discussed this with Jon today, and he said that it'd be useful to have a strict mode for such tests as well that we would enable during the BVT. By default, we could make the test general like this, so it would work on all setups, but during the BVT, it would fail if the suitable SR is not found. This would ensure that we catch errors like incorrect `allowed_operations` fields during the BVT. One way to implement this could be to have an optional SR parameter, and then during the BVT, we would pass the quicktest's SR to the function, and when this parameter isn't provided, the function could just run the test on all suitable SRs, if any. It's very useful in my opinion to have quicktests that work in all setups, because right now when I run the quicktests on my DT box, something always fails, and that's not useful for testing the quicktests.

To figure out how we could enable this strict quicktest mode during the BVT, we'd have to look at what the XenRT testcase corresponding to the quicktests does.